src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+
+    This class provides an interface to various optimization methods for models that
+    do not involve regularization. The optimization methods that are allowed for this
+    class are defined in the `allowed_optimizers` attribute.
+
+    Attributes
+    ----------
+    allowed_optimizers : list of str
+        List of optimizer names that are allowed for this solver class.
+
+    Methods
+    -------
+    instantiate_solver(loss)
+        Instantiates the optimization algorithm with the given loss function.
+
+    See Also
+    --------
+    [Solver](./#neurostatslib.solver.Solver) : Base solver class from which this class inherits.
+    """
+
+    allowed_optimizers = [
+        "GradientDescent",
+        "BFGS",
+        "LBFGS",
+        "ScipyMinimize",
+        "NonlinearCG",
+        "ScipyBoundedMinimize",
+        "LBFGSB",
+    ]
+
+    def __init__(
+        self, solver_name: str = "GradientDescent", solver_kwargs: Optional[dict] = None
+    ):
+        super().__init__(solver_name, solver_kwargs=solver_kwargs)
+
+    def instantiate_solver(
+        self,
+        loss: Callable[
+            [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jnp.ndarray
+        ],
+    ) -> Callable[
+        [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jaxopt.OptStep
+    ]:
+        """
+        Instantiate the optimization algorithm for a given loss function.
+
+        Parameters
+        ----------
+        loss :
+            The loss function that needs to be minimized.
+
+        Returns
+        -------
+        :
+            A runner function that uses the specified optimization algorithm
+            to minimize the given loss function.
+        """
+        self._check_is_callable_from_jax(loss)
+        solver_kwargs = self.solver_kwargs.copy()
+        solver_kwargs["fun"] = loss
+        return self.get_runner(solver_kwargs, {})
+
+
+class RidgeSolver(Solver):
+    """
+    Solver for Ridge regularization using various optimization algorithms.
+
+    This class uses `jaxopt` optimizers to perform Ridge regularization. It extends
+    the base Solver class, with the added feature of Ridge penalization.
+
+    Attributes
+    ----------
+    allowed_optimizers : List[..., str]
+        A list of optimizer names that are allowed to be used with this solver.
+    """
+
+    allowed_optimizers = [
+        "GradientDescent",
+        "BFGS",
+        "LBFGS",
+        "ScipyMinimize",
+        "NonlinearCG",
+        "ScipyBoundedMinimize",
+        "LBFGSB",
+    ]
+
+    def __init__(
+        self,
+        solver_name: str = "GradientDescent",
+        solver_kwargs: Optional[dict] = None,
+        regularizer_strength: float = 1.0,
+    ):
+        super().__init__(solver_name, solver_kwargs=solver_kwargs)
+        self.regularizer_strength = regularizer_strength
+
+    def penalization(self, params: Tuple[jnp.ndarray, jnp.ndarray]) -> jnp.ndarray:
```

Comment:
feels odd to give the ridge solver a public penalization method, but neither of the lasso solvers. do we think users should be accessing penalization directly? If not (which I think is the case), then this should be hidden

Reply:
yes, private would be better